custom_metrics/almanac.js

```diff
@@ -307,5 +335,248 @@ return JSON.stringify({
+      if (!img.hasAttribute('alt')) {
+        return -1;
+      }
+
+      // alt=" " is less correct but comparable to alt="" so we use trim
+      // Also remove duplicate spaces to get a feel for how long alt tags really are
+      return img.alt.trim().replace(/\s+/g, ' ').length;
+    });
+
+    return {
+      total_pictures: pictures.length,
+      total_img: img.length,
+      total_sources: sources.length,
+
+      total_with_srcset: images_with_srcset.length,
+      total_with_sizes: images_with_sizes.length,
+      total_pictures_with_img: pictures_with_img.length,
+
+      // Values specific properties. Cleaned and trimmed to make processing easier
+      sizes_values: images_with_sizes.map(img => img.sizes.replace(/[\r\n]+/g, ' ').replace(/\s+/g, ' ').trim()),
+      images_using_loading_prop: images_using_loading_prop.map(img => img.sizes.replace(/[\r\n]+/g, ' ').replace(/\s+/g, ' ').trim()),
+      alt_tag_lengths: alt_tag_lengths,
+
+      picture_props: parseNodes(pictures),
+      img_props: parseNodes(img),
+      source_props: parseNodes(sources),
+    };
+  }),
+
+  'videos': (() => {
+    const videos = document.querySelectorAll('video');
+
+    return {
+      total: videos.length,
+      props: parseNodes(videos),
+    };
+  }),
+
+  'scripts': (() => {
+    return {
+      total: document.scripts.length,
+      props: parseNodes(document.scripts),
+    };
+  }),
+
+  'nodes_using_role': (() => {
+    const nodes_with_role = [...document.querySelectorAll('[role]')];
+
+    /**
+     * 1. Build an object with each key being a unique value of `role` and the value being how often this role occurred
+     * 2. Make a list of unique role values
+     */
+    const unique_values = new Set();
+    const role_values_and_count = {};
+    for (const node of nodes_with_role) {
+      const role = node.getAttribute('role').toLowerCase();
+      unique_values.add(role);
+
+      if (!role_values_and_count[role]) {
+        role_values_and_count[role] = 1;
+        continue;
+      }
+
+      role_values_and_count[role]++;
+    }
+
+    return {
+      total: nodes_with_role.length,
+      values_and_count: role_values_and_count,
+      unique_values: [...unique_values],
+    };
+  }),
+
+  'total_nodes_with_duplicate_ids': (() => {
+    const nodes_with_id = [...document.querySelectorAll('[id]')];
+    const id_count_map = new Map();
+    for (const node of nodes_with_id) {
+      const count = id_count_map.get(node.id) || 0;
+      id_count_map.set(node.id, count + 1);
+    }
+
+    let total_duplicates = 0;
+    for (const count of id_count_map.values()) {
+      if (count > 1) {
+        // Subtract one because the first div to have this ID is not a duplicate
+        total_duplicates += count - 1;
+      }
+    }
+
+    return total_duplicates;
+  }),
+
+  /**
+   * The 'h' is stripped to create a numeric array.
+   * E.g. h1 > h2 > h3 > h3 => [1, 2, 3, 3, ]
+   */
+  'headings_order': (() => {
```

Comment:
I've a heading property that I could rename to headings and include this into.

Reply:
Sounds great. I'll leave that to you after this gets merged then :tada: